lektor/videotools.py

```diff
@@ -0,0 +1,332 @@
+        'width': None,
+        'height': None,
+        'duration': None,
+    }
+
+    # Try to extract total video duration
+    try:
+        info['duration'] = timedelta(
+            seconds=float(ffprobe_data['format']['duration']))
+    except (KeyError, TypeError, ValueError):
+        pass
+
+    # Try to extract width and height from the first found video stream
+    for stream in ffprobe_data['streams']:
+        if stream['codec_type'] != 'video':
+            continue
+
+        info['width'] = int(stream['width'])
+        info['height'] = int(stream['height'])
+
+        # We currently don't bother with multiple video streams
+        break
+
+    return info
+
+
+def make_video_thumbnail(ctx, source_video, source_url_path, seek, width=None,
+                         height=None, mode=ThumbnailMode.DEFAULT, upscale=None,
+                         quality=None, format=None):
+    if mode != ThumbnailMode.FIT and (width is None or height is None):
+        msg = '"%s" mode requires both `width` and `height` to be defined.'
+        raise ValueError(msg % mode.label)
+
+    if upscale is None:
+        upscale = {
+            ThumbnailMode.FIT: False,
+            ThumbnailMode.CROP: True,
+            ThumbnailMode.STRETCH: True,
+        }[mode]
+
+    if format is None:
+        format = 'jpg'
+    if format not in THUMBNAIL_FORMATS:
+        raise ValueError('Invalid thumbnail format "%s"' % format)
+
+    ffmpeg = locate_executable('ffmpeg')
+    if ffmpeg is None:
+        raise RuntimeError('Failed to locate ffmpeg')
+    info = get_video_info(source_video)
+
+    source_dim = Dimensions(info['width'], info['height'])
+    resize_dim, crop_dim = source_dim.resize(width, height, mode, upscale)
+
+    # Construct a filename suffix unique to the given parameters
+    suffix = get_suffix(seek, width, height, mode=mode, quality=quality)
+    dst_url_path = get_dependent_url(source_url_path, suffix,
+                                     ext='.{}'.format(format))
+
+    if quality is None:
+        quality = 95
+
+    @ctx.sub_artifact(artifact_name=dst_url_path, sources=[source_video])
+    def build_thumbnail_artifact(artifact):
+        artifact.ensure_dir()
+
+        vfilter = 'thumbnail,scale={rw}:{rh},crop={tw}:{th}'.format(
+            rw=resize_dim.width,
+            rh=resize_dim.height,
+            tw=crop_dim.width,
+            th=crop_dim.height,
+        )
+
+        cmdline = [
+            ffmpeg,
+            '-loglevel', '-8',
+            '-ss', get_timecode(seek),  # Input seeking since it's faster
+            '-i', source_video,
+            '-vf', vfilter,
+            '-frames:v', '1',
+            '-qscale:v', str(get_ffmpeg_quality(quality)),
+            artifact.dst_filename,
+        ]
+
+        reporter.report_debug_info('ffmpeg cmd line', cmdline)
+        proc = portable_popen(cmdline)
+        if proc.wait() != 0:
+            raise RuntimeError('ffmpeg exited with code %d' % proc.returncode)
```

Comment:
I think using an exception is best for now. If someone provides a compelling use case we can always allow it later. I think it's better to be a bit too restrictive in the beginning. `min(seek, video.duration)` can be used as a workaround if necessary.

Reply:
In these sorts of cases, I like to think about what the user intends when they specify seek > duration. I don't see it as likely that they meant "seek to the end and produce a 0-frame file", so an exception sounds good to me too. Does anyone else see an alternate use case?

Reply:
Let's go with an exception if seek > duration, and if duration is indeterminate have a quick check for a output image really existing with another exception if it doesn't. It was actually no file, not an empty file / bad contents. I don't know how likely it will be that we can't determine a duration.

The only use case I can contrive is if you for some reason are seeking the same thing on many files, and that it's valid for some/most. But I agree, we can loosen it later if there's a good reason to.

Reply:
I ended up throwing an exception if a file was not created by ffmpeg, with a suggestion that it may be outside of the duration. There is also validation in `VideoFrame` to ensure we don't end up in those cases.